### PR TITLE
feat: add retry logic to assume role

### PR DESCRIPTION
**Pull request by blz-ea**
*Issue #, if available:*
#299 

*Description of changes:*
Add retry logic to assume role function

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -236,6 +239,10 @@ function getStsClient(region) {
   });
 }
 
+function isRetryableError(error) {
+  return error instanceof Error && error.code && (error.code === 'IDPCommunicationErrorException' || error.code === 'InvalidIdentityToken');
```
Comment (miluoshi) [inline index.js:243]: Can it be simplified to this?

```js
return error instanceof Error && ['IDPCommunicationErrorException', 'InvalidIdentityToken'].includes(error?.code);
```

Comment (blz-ea) [reply index.js:243]: This is nice

File: action.yml [hunk 2]
```diff
@@ -51,6 +51,12 @@ inputs:
   role-skip-session-tagging:
     description: 'Skip session tagging during role assumption'
     required: false
+  role-to-assume-retries:
+    description: 'Number of times assume role will be re-tried (default: 5)'
+    required: false
+  role-to-assume-retries-factor:
+    description: 'The exponential factor to use for retries. (default: 5)'
```
[the remark this answers is not in the document]
Comment (blz-ea) [reply action.yml:58]: Oh

File: index.js (hunk 1 above)
Comment (blz-ea) [reply index.js]: @miluoshi But build fails with
```
You may need an additional loader to handle the result of these loaders.
| 
| function isRetryableError(error) {
>   return error instanceof Error && ['IDPCommunicationErrorException', 'InvalidIdentityToken'].includes(error?.code);
| }

```

Comment (blz-ea) [reply index.js]: So probably just `return error instanceof Error && ['IDPCommunicationErrorException', 'InvalidIdentityToken'].includes(error.code);` should work

**Comment by miluoshi** [comment]
Any update on this? Should we tag someone from the aws-actions org to look at it?

**Comment by GarethLoweCinch** [comment]
It would be great to see this merged - we are getting impacted in production by OIDC failures as a result of this.

Tagging @mattsb42-aws and @clareliguori

**Comment by paragbhingre** [comment]
> It would be great to see this merged - we are getting impacted in production by OIDC failures as a result of this.
> 
> Tagging @mattsb42-aws and @clareliguori

Hi @GarethLoweCinch , We (ECS Team) are looking into the issue and we were able to reproduce the issue on our side. We are investigating on the solution right now, so we will keep you posted about it. 
FYI @efekarakus

**Comment by pawelmitka** [comment]
We are facing this issue as well. It would be great to have this PR merged or the issue fixed another way. 
FYI: @winglot

**Comment by yvio** [comment]
Envoyé de mon iPhone

**Comment by dannyburke1** [comment]
Hello. I've forked this PR and released it here: 
`dannyburke1/configure-aws-credentials@v1.6.0` - I'm currently battle testing it, but so far seems to be working OK.

**Comment by dannyburke1** [comment]
Hi there, 

I've published this version on `dannyburke1/configure-aws-credentials@v1.7.2` and set 10 parallel jobs to assume. Typically, this fails on one in 1 in 10 parallel builds, so it this was the best way to test it - however I'm not sure this is 100% retrying the call as I'm still getting failed builds even when setting the retry count to something ridiculous like 500, the error is returned and handled straight away rather than retrying and taking a few minutes- at least this is my experience. I've also tried catching some additional errors to see if that helped but unfortunately didn't. Happy to work with you and show you my changes in order to get this over the fence.

Cheers,
danny

**Comment by blz-ea** [comment]
Interesting [comment here](https://github.com/actions/runner/issues/1466#issuecomment-1007220359). So it seems Github is not able to respond it time

**Comment by blz-ea** [comment]
@dannyburke1 Currently retrying limited to `IDPCommunicationError` and `InvalidIdentityToken` errors (from [this list](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithWebIdentity.html#API_AssumeRoleWithWebIdentity_Errors)). It would be nice if you could share error codes/messages that you've encountered

**Pull request closed without merging**